twitter/src/main/java/com/hazelcast/jet/contrib/twitter/TwitterSources.java

```diff
@@ -0,0 +1,390 @@
+        Objects.requireNonNull(key, description);
+        if ("REPLACE_THIS".equals(key)) {
+            throw new IllegalArgumentException("Twitter credentials key '" + description
+                    + "' contains the default placeholder");
+        }
+    }
+
+    /**
+     * A context for the stream source of Twitter's Stream API
+     */
+    private static final class TwitterStreamSourceContext {
+
+        private static final int QUEUE_CAPACITY = 1000;
+        private static final int MAX_FILL_ELEMENTS = 250;
+
+        private final BlockingQueue<String> queue = new LinkedBlockingQueue<>(QUEUE_CAPACITY);
+        private final ArrayList<String> buffer = new ArrayList<>(MAX_FILL_ELEMENTS);
+        private final BasicClient client;
+
+        /**
+         * @param credentials      a Twitter OAuth1 credentials that consists
+         *                         "consumerKey",  "consumerSecret", "token",
+         *                         "tokenSecret" keys.
+         * @param host             a Twitter host URL to connect. These hosts
+         *                         are defined in {@link com.twitter.hbc.core.Constants}.
+         * @param endpointSupplier Supplier that supplies a Twitter
+         *                         StreamingEndpoint to connect to source.
+         */
+        private TwitterStreamSourceContext(
+                @Nonnull Properties credentials,
+                @Nonnull String host,
+                @Nonnull SupplierEx<? extends StreamingEndpoint> endpointSupplier
+        ) {
+            checkTwitterCredentials(credentials);
+            String consumerKey = credentials.getProperty("consumerKey");
+            String consumerSecret = credentials.getProperty("consumerSecret");
+            String token = credentials.getProperty("token");
+            String tokenSecret = credentials.getProperty("tokenSecret");
+
+            Authentication auth = new OAuth1(consumerKey, consumerSecret, token, tokenSecret);
+            client = new ClientBuilder()
+                    .hosts(host)
+                    .endpoint(endpointSupplier.get())
+                    .authentication(auth)
+                    .processor(new StringDelimitedProcessor(queue))
+                    .build();
+            client.connect();
+        }
+
+        private void fillBuffer(SourceBuilder.SourceBuffer<String> sourceBuffer) {
+            queue.drainTo(buffer, MAX_FILL_ELEMENTS);
+            for (String item : buffer) {
+                sourceBuffer.add(item);
+            }
+            buffer.clear();
+        }
+
+        private void fillTimestampedBuffer(SourceBuilder.TimestampedSourceBuffer<String> sourceBuffer) {
+            queue.drainTo(buffer, MAX_FILL_ELEMENTS);
+            for (String item : buffer) {
+                JsonObject object = Json.parse(item).asObject();
+                long timestamp = Long.parseLong(object.getString("timestamp_ms", "0"));
```

Comment:
A better default might be `Long.MIN_VALUE`. The `0` will actually be used as a time and if there's a windowing processor downstream, the message will be thrown out as late anyway (unless your system clock is before year 1970, in that case it would cause all other messages to be late).

I'm also not sure if we shouldn't ignore malformed messages here, maybe just log them. I saw it done in Twitter's own code in `StringDelimitedProcessor.processNextMessage()`. You can get the logger in `Context` passed to the SourceBuilder's `createFn`.

Reply:
I think this issue is resolved after your corrections.